README.md
# Impact of seed selection on subclonal reconstruction solutions

Project goal:
1. Quantify SRC result variability across SRC tools
2. Quantify SRC result variability across initializing random seed


## Description

In this project, we have expanded the subclonal reconstruction pipeline https://github.com/uclahs-cds/pipeline-call-SRC to accept the output of multiple mutation callers.

We have integrated 4 additional mutation callers (MuTect2, SomaticSniper, Strelka2, Battenberg) by creating parsers that extract variant data from the different tools' output. The parsers can be found here https://github.com/uclahs-cds/tool-SRC-util.   

## Random Seed Selection

### Generating random seeds:

Generate initial seed:
`head -c 4 /dev/urandom | od -An -tu4` => `3058353505`

Generate 10 random seeds:

```
import random
random.seed(3058353505)
random.sample(range(0, 1000000), k=10)
```

### Chosen random seeds:

Generated/chosen seeds:
`[51404, 366306, 423647, 838004, 50135, 628019, 97782, 253505, 659767, 13142]`

## Tumour Sample Data:

[HNSC] Head and Neck cohort: `/hot/project/disease/HeadNeckTumor/HNSC-000084-LNMEvolution/data/`

[CPCGENE] Prostate cohort: `/hot/project/disease/ProstateTumor/PRAD-000005-293PT/`

## Tools:

sSNV-caller: `mutect2`, `strelka2`, `somaticsniper`

sCNA-caller: `battenberg`

src-tool: `pyclone-vi`, `dpclust`, `phylowgs` 

### Node selection:
- PyClone-VI: F32 (average run time 15s - 10min)
- DPClust: F32 (average run time 5min - 40min)
- PhyloWGS: F72 (average run time 7h - 24h)

## sSNV-caller and sCNA-caller results:

sSNV-caller output: 

`/hot/project/disease/HeadNeckTumor/HNSC-000084-LNMEvolution/data/SNV/<sSNV-caller>/recsnv/vcfs/` 

sCNA-caller output: 

`/hot/project/disease/HeadNeckTumor/HNSC-000084-LNMEvolution/data/<sCNA-caller>/`


## Modes:
[HNSC]
- Single region mode (sr) (run on primary tumour only)
- Multi region mode (mr) (run on primary and metastatic tumours)

[CPGENE]
- Multi region mode (mr) (run on multiple regions of primary tumour)

## Running The Pipeline

The pipeline is run for each sample and seed in both single region and multi region mode.

Templates: `/hot/project/method/AlgorithmEvaluation/BNCH-000082-SRCRNDSeed/pipeline-call-src/templates/`

### Pipeline input files:
configs: 

`/hot/project/method/AlgorithmEvaluation/BNCH-000082-SRCRNDSeed/pipeline-call-src/run-<sSNV-caller>-<sCNA-caller>-<src-tool>/input/config/seed_<seed>.config`

- 1 config per `seed`
- indicates `src-tool` choice and pipeline parameters
- indicates pipeline run output directory

yamls: 
 
`/hot/project/method/AlgorithmEvaluation/BNCH-000082-SRCRNDSeed/pipeline-call-src/<sSNV-caller>_<sCNA-caller>_yamls/`

single-region mode (one primary tumor sample:
`sinlge-region/<sample_id>.yaml`

multi-region mode (primary and metastatic tumor samples):
`multi-region/<patient_id>.yaml`

- 1 yaml per patient 
- path to `sSNV-caller` output
- path to `sCNA-caller` output

### Submission script:

Log files:

`/hot/project/method/AlgorithmEvaluation/BNCH-000082-SRCRNDSeed/pipeline-call-src/run-<sSNV-caller>-<sCNA-caller>-<src-tool>/logs/`

Create conda environment with `pyyaml` and `numpy` installed. Execute submission script in logs directory and from activated conda environment. 

submission script: `<mode>_<sSNV-caller>_<sCNA-caller>_<src-tool>_submission_script.sh`

### Completed pipeline combinations:

- [x] Strelka2-Battenberg-PyClone-VI (sr/mr)
- [x] Strelka2-Battenberg-DPClust (sr)
- [x] Strelka2-Battenberg-PhyloWGS (sr)
- [x] SomaticSniper-Battenberg-PyClone-VI (sr/mr)
- [x] SomaticSniper-Battenberg-DPClust (sr)
- [x] SomaticSniper-Battenberg-PhyloWGS (sr)
- [x] Mutect2-Battenberg-PyClone-VI (sr/mr)
- [x] Mutect2-Battenberg-DPClust (sr)
- [x] Mutect2-Battenberg-PhyloWGS (sr)

<img width="875" alt="Screen Shot 2023-05-30 at 8 05 47 AM" src="https://github.com/uclahs-cds/project-method-AlgorithmEvaluation-BNCH-000082-SRCRNDSeed/assets/108153291/7b3e7dbc-a56d-4cfa-9f11-38b15d0688c9">

### Results:

Output files:

`/hot/project/method/AlgorithmEvaluation/BNCH-000082-SRCRNDSeed/pipeline-call-src/run-<sSNV-caller>-<sCNA-caller>-<src-tool>/output/`

## License

Author: Anna Neiman-Golden(aneimangolden@mednet.ucla.edu), Philippa Steinberg(psteinberg@mednet.ucla.edu)

[This project] is licensed under the GNU General Public License version 2. See the file LICENSE.md for the terms of the GNU GPL license.

<one line to give the project/program's name and a brief idea of what it does.>

Copyright (C) 2021 University of California Los Angeles ("Boutros Lab") All rights reserved.

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.


CHANGELOG.md
# Changelog
All notable changes to the pipeline running and analysis of *project-AlgorithmEvaluation-BNCH-000082-SRCRNDSeed*.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

---

## [Unreleased]
### Changed
- Changed something but it is not part of the last release.

---

## [0.0.1] - YYYY-MM-DD
### Added
- For new analysis or config files.
- Added test script for pipeline run using Mutect2.
- Added test script for pipeline run using Strelka2.
- Added test script for pipeline run using SomaticSniper.
- Added script to run Strelka2-Battenberg-Pyclone-VI pipeline single-sample mode.
- Added script to run Strelka2-Battenberg-Pyclone-VI pipeline multi-sample mode.
- Added script to run Strelka2-Battenberg-DPClust pipeline single-sample mode.
- Added script to parse patient, seed and number of subclones from Strelka2-Battenberg-PyClone-VI pipeline single-sample mode output.
- Added script to parse patient, seed and number of subclones from Strelka2-Battenberg-DPClust pipeline single-sample mode output.
- Added script to parse patient, seed and number of subclones from Strelka2-Battenberg-PyClone-VI pipeline multi-sample mode output.
- Added script for scatterplots of relative seed variability for DPClust, PyClone-VI single-sample and multi-sample mode.

### Changed
- For changes in existing config files, commands and analyses.
- Changed item 1.
- Changed README to include changed file directories and project details. 

### Deprecated
- For soon-to-be removed out-of-date configs and analyses.

### Removed
- For now removed configs and analyses.  

### Fixed
- For any bug fixes in existing config files, commands and analyses.
- Fixed item 1.

### Security
- In case of vulnerabilities.

## [Unreleased] - 2022-11-16
### Added
- new script sr_strelka2_battenberg_dpclust_submission_script.sh
- new script sr_strelka2_battenberg_pyclone-vi_submission_script.sh

### Removed
- Old script ss_strelka2_battenberg_dpclust_submission_script.sh

## [Unreleased] - 2022-11-28
### Added
- Added new script sr_somaticsniper_battenberg_pyclone-vi_submission_script.sh
- Added new script sr_somaticsniper_battenberg_dpclust_submission_script.sh
- Added new script mr_strelka2_battenberg_pyclone-vi_submission_script.sh
- Added new script sr_mutect2_battenberg_dpclust_submission_script.sh

### Removed
- Removed old script ms_strelka2_battenberg_pyclone_submission_script.sh

## [Unreleased] - 2022-12-04
### Added
- Added new script sr_mutect2_battenberg_pyclone-vi_submission_script.sh

## [Unreleased] - 2022-12-05
### Added
- Added new script mr_somaticsniper_battenberg_pyclone-vi_submission_script.sh
- Added new script mr_mutect2_battenberg_pyclone-vi_submission_script.sh

### Removed
- Removed old script sr_strelka2_battenberg_pyclone-vi_submission_script.sh

## [Unreleased] - 2023-03-07
### Added
- Added new script sr_mutect2_battenberg_phylowgs_submission_script.sh

## [Unreleased] - 2023-04-13
### Added
- Added new script sr_strelka2_battenberg_phylowgs_submission_script.sh

## [Unreleased] - 2023-04-17
### Added
- Added new plotting script subclone_relative_seed_variability.R

## [Unreleased] - 2023-04-23
### Added
- Added new parsing script parse_num_subclones_phyloWGS.R
- Added new plotting script plot_num_subclones_box.R
- Added plot_accessories.R

### Changed
- Moved seed, colour scheme and legends to plot_accesories.R

## [Unreleased] - 2023-04-27
### Added
- Added script for running parse_num_subclones_phyloWGS.R

### Changed
- Split phyloWGS parsing into parse_consensus_tree_phyloWGS.R and parse_num_subclones_phyloWGS.R

## [Unreleased] - 2023-05-08
### Added
- Added new script sr_somaticsniper_battenberg_phylowgs_submission_script.sh

## [Unreleased] - 2023-08-22
### Added
- Added new analysis script supplementary_tables_stats.R


src/output-analysis/output_parser.R
### PARSER ########################################################################################
# Extract Strelka2-Battenberg-PyClone-VI output from 14 head and neck samples across 10 seeds.

### PREAMBLE ######################################################################################
# Input directory stem to all pipeline output files
input.dir.stem <- '/hot/project/method/AlgorithmEvaluation/BNCH-000082-SRCRNDSeed/pipeline-call-src/'
input.pipeline <- 'run-strelka2-battenberg-pyclone-vi/output/pipeline-call-SRC-1.0.0-rc.1/%s/PyClone-VI-0.1.2/output/'
input.file <- 'PyClone-VI-0.1.2_%s_%s_Strelka2-Battenberg.tsv'

# Output directory
output.dir.stem <- '/hot/project/method/AlgorithmEvaluation/BNCH-000082-SRCRNDSeed/'
output.pipeline <- 'pipeline-call-src/run-strelka2-battenberg-pyclone-vi/output/'
output.file <- 'all_subclones_per_patient_seed.tsv'

# Pipelines run
pipeline.runs <- c(
  'run-strelka2-battenberg-pyclone-vi',
  'run-strelka2-battenberg-dpclust'
  );

# 10 random seeds and 14 primary tumour sample data.
seeds <- c(
    '51404',
    '366306',
    '423647',
    '838004',
    '50135',
    '628019',
    '97782',
    '253505',
    '659767',
    '13142'
    );

samples <- c(
    'ILHNLNEV000001-T001-P01-F',
    'ILHNLNEV000002-T001-P01-F',
    'ILHNLNEV000003-T001-P01-F',
    'ILHNLNEV000004-T001-P01-F',
    'ILHNLNEV000005-T001-P01-F',
    'ILHNLNEV000006-T001-P01-F',
    'ILHNLNEV000007-T001-P01-F',
    'ILHNLNEV000008-T001-P01-F',
    'ILHNLNEV000009-T001-P01-F',
    'ILHNLNEV000010-T001-P01-F',
    'ILHNLNEV000011-T001-P01-F',
    'ILHNLNEV000012-T001-P01-F',
    'ILHNLNEV000013-T001-P01-F',
    'ILHNLNEV000014-T001-P01-F'
    );
### CODE #########################################################################################
# Get all output paths (Anna)
all.paths <- list();
for (i in 1:length(samples)){
    all.paths[i] <- list(rep(NA, length(seeds)))
    };

for (sample in 1:length(samples)) {
    for (seed in 1:length(seeds)) {
        path <- sprintf(
            fmt = paste0(input.dir.stem, input.pipeline, input.file),
            samples[sample],
            seeds[seed],
            samples[sample]
        )
        all.paths[[sample]][seed] <- path
        };
    };

### get.patient.seed.summary ######################################################################
# Get summary data per patient file without creating lists of tables
get.patient.seed.summary <- function(file.path) {
    # read in file
    table <- read.table(
        file = file.path,
        sep = '\t',
        header = TRUE
        );
    seed <- strsplit(file.path, '_')[[1]][2]
    sample <- strsplit(file.path, '_')[[1]][3]
    patient <- strsplit(sample, '-')[[1]][1]
    subclones <- length(unique(table$cluster_id))
    sample.summary <- list(patient, seed, subclones)
    return(sample.summary)
    };

### write.table ###################################################################################
# Add patient vs. seed summary data to summary table
all.samples.summary <- data.frame();
for (paths in all.paths) {
    for (path in paths) {
        all.samples.summary <- rbind(all.samples.summary, get.patient.seed.summary(path))
        };
    };
colnames(all.samples.summary) <- c('patient', 'seed', 'n_clones');
write.table(
    x = all.samples.summary,
    file = paste0(output.dir.stem, output.pipeline, Sys.Date(), '_', output.file),
    sep = '\t',
    quote = FALSE,
    row.names = FALSE
    );
